message.go

```diff
@@ -168,13 +167,56 @@ type MessageReactions struct {
 
 // ContentWithMentionsReplaced will replace all @<id> mentions with the
 // username of the mention.
-func (m *Message) ContentWithMentionsReplaced() string {
-	if m.Mentions == nil {
-		return m.Content
+func (m *Message) ContentWithMentionsReplaced() (content string) {
+	content = m.Content
+
+	for _, user := range m.Mentions {
+		content = regexp.MustCompile("<@!?"+regexp.QuoteMeta(user.ID)+">").ReplaceAllString(content, "@"+user.Username)
+	}
+	return
+}
+
+var patternRoles *regexp.Regexp
+
+// ContentWithMoreMentionsReplaced will replace all @<id> mentions with the
+// username of the mention, but also role IDs and more.
+func (m *Message) ContentWithMoreMentionsReplaced(s *Session) (content string, err error) {
+	content = m.Content
+
+	if !s.StateEnabled {
+		content = m.ContentWithMentionsReplaced()
+		return
 	}
-	content := m.Content
+
+	channel, err := s.State.Channel(m.ChannelID)
+	if err != nil {
+		return
+	}
+
 	for _, user := range m.Mentions {
-		content = regexp.MustCompile(fmt.Sprintf("<@!?(%s)>", user.ID)).ReplaceAllString(content, "@"+user.Username)
+		member, err := s.State.Member(channel.GuildID, user.ID)
+		if err != nil {
+			continue
+		}
+
+		nick := member.Nick
+		if nick == "" {
+			nick = user.Username
+		}
+		content = regexp.MustCompile("<@!?"+regexp.QuoteMeta(user.ID)+">").ReplaceAllString(content, "@"+nick)
 	}
-	return content
+
+	if patternRoles == nil {
+		patternRoles = regexp.MustCompile("<&[^>]*>")
+	}
+
+	patternRoles.ReplaceAllStringFunc(content, func(mention string) string {
+		role, err := s.State.Role(channel.GuildID, mention)
```

Comment:
This won't work right? You're passing `<&40394820394209>` here as the role id instead of just `40394820394209`, in bruxism I just do mention[3 : len(mention)-1]